daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala

```diff
@@ -298,231 +299,217 @@ class Runner(
+                val t = Timestamp.assertFromInstant(timeProvider.getCurrentTime)
+                run(SEApp(SEValue(continue), Array(SEValue(STimestamp(t)))))
+              }
+              case SVariant(_, "Sleep", _, v) => {
+                v match {
+                  case SRecord(_, _, vals) if vals.size == 2 => {
+                    val continue = vals.get(1)
+                    for {
+                      sleepMicros <- vals.get(0) match {
+                        case SRecord(_, _, vals) if vals.size == 1 =>
+                          vals.get(0) match {
+                            case SInt64(i) => Future.successful(i)
+                            case _ =>
+                              Future.failed(new ConverterException(s"Expected SInt64 but got $v"))
+                          }
+                        case v =>
+                          Future.failed(new ConverterException(s"Expected RelTime but got $v"))
+                      }
+                      v <- {
+                        val sleepMillis = sleepMicros / 1000
+                        val sleepNanos = (sleepMicros % 1000) * 1000
+                        Thread.sleep(sleepMillis, sleepNanos.toInt)
+                        run(SEApp(SEValue(continue), Array(SEValue(SUnit))))
+                      }
+                    } yield v
                   }
+                  case _ =>
+                    Future.failed(
+                      new ConverterException(s"Expected record with 2 fields but got $v"))
                 }
-                case _ =>
-                  Future.failed(
-                    new ConverterException(s"Expected Submit, Query or AllocParty but got $v"))
               }
+              case _ =>
+                Future.failed(
+                  new ConverterException(s"Expected Submit, Query or AllocParty but got $v"))
             }
-            case SVariant(_, "Pure", _, v) =>
-              v match {
-                case SRecord(_, _, vals) if vals.size == 2 => {
-                  // Unwrap the Tuple2 we get from the inlined StateT.
-                  Future { vals.get(0) }
-                }
-                case _ => Future.failed(new ConverterException(s"Expected Tuple2 but got $v"))
+          }
+          case SVariant(_, "Pure", _, v) =>
+            v match {
+              case SRecord(_, _, vals) if vals.size == 2 => {
+                // Unwrap the Tuple2 we get from the inlined StateT.
+                Future { vals.get(0) }
               }
-            case v => Future.failed(new ConverterException(s"Expected Free or Pure but got $v"))
-        })
+              case _ => Future.failed(new ConverterException(s"Expected Tuple2 but got $v"))
+            }
+          case v => Future.failed(new ConverterException(s"Expected Free or Pure but got $v"))
+        }
     }
 
     for {
-      _ <- stepToValue().fold(Future.failed(_), Future.successful(_))
-      _ <- machine.toSValue match {
+      _ <- Future.unit
```

Comment:
You probably want the first step to be run in a Future, don't you ?

Reply:
Makes sense I guess but please add a comment.